The government of Kenya would like to improve the state of sanitation in primary schools all over the country.  We have been approached as a Data Science team to analyze  Kenyan primary schools datasets to identify the schools, provinces, districts and counties that need to upgrade their sanitation facilities.

# Importing Our Libraries

In [ ]:
# Importing the Pandas Library
import pandas as pd

# Importing the Numpy Library
import numpy as np

In [ ]:
# from google.colab import drive
# drive.mount('/content/drive')

# Loading our Dataset

In [ ]:
# Loading our counties dataset first
counties=pd.read_csv('/content/Counties.csv')
counties

district       county
0       BARINGO      BARINGO
1         BOMET      KERICHO
2         BONDO        SIAYA
3       BUNGOMA      BUNGOMA
4         BURET      KERICHO
..          ...          ...
66      TURKANA      TURKANA
67  UASIN GISHU  UASIN GISHU
68       VIHIGA       VIHIGA
69        WAJIR        WAJIR
70   WEST POKOT   WEST POKOT

[71 rows x 2 columns]

In [ ]:
# let us now load the dataset we will be working on
# And preview the first 5 rows of our dataset
df = pd.read_csv('/content/kenya_primary_schools.csv')
df.head()

X     Y    FID  ... Costituenc Latitude Longitude
0  36.14  0.27  10001  ...    SINENDE     0.27     36.14
1  36.17  0.26  10002  ...    SINENDE     0.26     36.17
2  36.18  0.24  10003  ...    KABUSWO     0.24     36.18
3  36.19  0.21  10004  ...    KABUSWO     0.21     36.19
4  37.70 -0.88  10005  ...    MASINGA    -0.88     37.70

[5 rows x 38 columns]

# Getting Information about our Dataset

In [ ]:
# Finding out information about our counties dataset
counties.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   district  71 non-null     object
 1   county    71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [ ]:
# Finding out information about our primary school dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31230 entries, 0 to 31229
Data columns (total 38 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   X            31230 non-null  float64
 1   Y            31230 non-null  float64
 2   FID          31230 non-null  int64  
 3   Name_of_Sc   31229 non-null  object 
 4   Level_       31229 non-null  object 
 5   Status       31229 non-null  object 
 6   SchSponsor   31229 non-null  object 
 7   Type1        31229 non-null  object 
 8   Type2        31229 non-null  object 
 9   Type3        31229 non-null  object 
 10  PupilTeach   31229 non-null  float64
 11  ClassrmRat   31229 non-null  float64
 12  ToiletRati   31229 non-null  float64
 13  No_Classrm   31229 non-null  float64
 14  BoysToilet   31229 non-null  float64
 15  GirlsToilet  31229 non-null  float64
 16  TeachersTo   31229 non-null  float64
 17  TotalToile   31229 non-null  float64
 18  TotalBoys    31229 non-null  float64
 19  Tota

# Data Cleaning

Lets clean our primary school dataset

### Validity

In [ ]:
# Checking for similarity of columns
df[ ~(df.X == df['Longitude'])]

X         Y    FID  ... Costituenc Latitude Longitude
9826   39.599613 -3.790685   1827  ...   KALOLENI    -3.79     39.59
24229  37.840855 -0.492438  31230  ...        NaN      NaN       NaN

[2 rows x 38 columns]

In [ ]:
# Dropping unnecesary columns
prim = df.drop(['X','Y','Level_','ClassrmRat','PupilTeach',\
         'GO_KTSC_M','GOK_TSC_F','TotalBoys','TotalGirls','authorityM','authorityF',\
         'PTA_BOG_M','PTA_BOG_F','OthersM','OthersF','NoTeaching','NonTeachin',\
         'Division','Latitude','Longitude', 'SchSponsor'], axis = 1)
prim.head()

FID          Name_of_Sc  Status  ...  District Location Costituenc
0  10001      CHEBIREBEI PRI  PUBLIC  ...  KOIBATEK  MOGOTIO    SINENDE
1  10002         WASEGES PRI  PUBLIC  ...  KOIBATEK  MOGOTIO    SINENDE
2  10003         KIRIBOT PRI  PUBLIC  ...  KOIBATEK  MOGOTIO    KABUSWO
3  10004     KABUSWO PRI SCH  PUBLIC  ...  KOIBATEK  MOGOTIO    KABUSWO
4  10005  VONDENI PRI SCHOOL  PUBLIC  ...  MACHAKOS    KIVAA    MASINGA

[5 rows x 17 columns]

### Completeness

In [ ]:
# Check for null values
prim.isnull().sum()

FID            0
Name_of_Sc     1
Status         1
Type1          1
Type2          1
Type3          1
ToiletRati     1
No_Classrm     1
BoysToilet     1
GirlsToilet    1
TeachersTo     1
TotalToile     1
TotalEnrol     1
Province       1
District       1
Location       1
Costituenc     1
dtype: int64

In [ ]:
# Dropping null values
prim_1 = prim.dropna()

### Consistency

In [ ]:
# Checking for duplicates
prim_1.duplicated().sum()

0

### Uniformity

In [ ]:
pd.options.mode.chained_assignment = None  

In [ ]:
# Changing the case of column names to lower, stripping whitespaces,
# removing periods and replacing spaces with underscore.
prim_1.columns = prim_1.columns.str.lower().str.strip().str.replace(' ', '_')
prim_1.head()

fid          name_of_sc  status  ...  district location costituenc
0  10001      CHEBIREBEI PRI  PUBLIC  ...  KOIBATEK  MOGOTIO    SINENDE
1  10002         WASEGES PRI  PUBLIC  ...  KOIBATEK  MOGOTIO    SINENDE
2  10003         KIRIBOT PRI  PUBLIC  ...  KOIBATEK  MOGOTIO    KABUSWO
3  10004     KABUSWO PRI SCH  PUBLIC  ...  KOIBATEK  MOGOTIO    KABUSWO
4  10005  VONDENI PRI SCHOOL  PUBLIC  ...  MACHAKOS    KIVAA    MASINGA

[5 rows x 17 columns]

In [ ]:
# Renaming columns to have more relevant names
prim_1.rename(columns = {'name_of_sc':'school_name', 'schsponsor': 'school_sponsor','totaltoile':'total_toilets',\
                         'toiletrati':'toilet_ratio','totalenrol':'total_students','teachersto':'teachers_toilet',\
                         'costituenc':'constituency'}, inplace = True)
prim_1.head()

fid         school_name  status  ...  district location constituency
0  10001      CHEBIREBEI PRI  PUBLIC  ...  KOIBATEK  MOGOTIO      SINENDE
1  10002         WASEGES PRI  PUBLIC  ...  KOIBATEK  MOGOTIO      SINENDE
2  10003         KIRIBOT PRI  PUBLIC  ...  KOIBATEK  MOGOTIO      KABUSWO
3  10004     KABUSWO PRI SCH  PUBLIC  ...  KOIBATEK  MOGOTIO      KABUSWO
4  10005  VONDENI PRI SCHOOL  PUBLIC  ...  MACHAKOS    KIVAA      MASINGA

[5 rows x 17 columns]

In [ ]:
# Filling totaltoilet column with sum of boystoilet and girlstoilet
toilets = prim_1['boystoilet']+prim_1['girlstoilet']
prim_1['total_toilets']= toilets
prim_1

fid         school_name  status  ...  district  location   constituency
0      10001      CHEBIREBEI PRI  PUBLIC  ...  KOIBATEK   MOGOTIO        SINENDE
1      10002         WASEGES PRI  PUBLIC  ...  KOIBATEK   MOGOTIO        SINENDE
2      10003         KIRIBOT PRI  PUBLIC  ...  KOIBATEK   MOGOTIO        KABUSWO
3      10004     KABUSWO PRI SCH  PUBLIC  ...  KOIBATEK   MOGOTIO        KABUSWO
4      10005  VONDENI PRI SCHOOL  PUBLIC  ...  MACHAKOS     KIVAA        MASINGA
...      ...                 ...     ...  ...       ...       ...            ...
31225  30996        MGANGANI PRI  PUBLIC  ...  MACHAKOS    YATHUI          MWALA
31226  30997        KISIKINI PRI  PUBLIC  ...     KITUI  MBUSYANI  KITUI CENTRAL
31227  30998     NGOMANO PRI SCH  PUBLIC  ...  MACHAKOS    YATHUI          MWALA
31228  30999         NGILUNI PRI  PUBLIC  ...     KITUI  MBUSYANI  KITUI CENTRAL
31229  31000      KIKASO PRI SCH  PUBLIC  ...  MACHAKOS    YATHUI          MWALA

[31229 rows x 17 columns]

In [ ]:
# Renaming of Central province to CENTRAL for uniformity.
prim_1['province'].replace({'Central':'CENTRAL'},inplace=True)

### Merging our datasets

Now that we have cleaned up our primary school dataset, lets merge it to the counties dataset.

In [ ]:
# Merging our two datasets
new_df = prim_1.merge(counties, on= ['district'], how= 'outer')
new_df

fid       school_name  status  ... location constituency    county
0      10001.0    CHEBIREBEI PRI  PUBLIC  ...  MOGOTIO      SINENDE   BARINGO
1      10002.0       WASEGES PRI  PUBLIC  ...  MOGOTIO      SINENDE   BARINGO
2      10003.0       KIRIBOT PRI  PUBLIC  ...  MOGOTIO      KABUSWO   BARINGO
3      10004.0   KABUSWO PRI SCH  PUBLIC  ...  MOGOTIO      KABUSWO   BARINGO
4      10006.0         POMBO PRI  PUBLIC  ...  MOGOTIO    NGENDALEL   BARINGO
...        ...               ...     ...  ...      ...          ...       ...
31225  29447.0    URAN GODHA PRI  PUBLIC  ...     URAN       MOYALE  MARSABIT
31226  29451.0         WALDA PRI  PUBLIC  ...    WALDA       MOYALE  MARSABIT
31227  29455.0  DADACH ELELE PRI  PUBLIC  ...    WALDA       MOYALE  MARSABIT
31228  29458.0        RAWANA PRI  PUBLIC  ...   RAWANA       MOYALE  MARSABIT
31229      NaN               NaN     NaN  ...      NaN          NaN   KERICHO

[31230 rows x 18 columns]

### Cleaning after Merging

In [ ]:
# Checking for null values in our merged dataframe
new_df.isnull().sum()

fid                  1
school_name          1
status               1
type1                1
type2                1
type3                1
toilet_ratio         1
no_classrm           1
boystoilet           1
girlstoilet          1
teachers_toilet      1
total_toilets        1
total_students       1
province             1
district             0
location             1
constituency         1
county             313
dtype: int64

In [ ]:
# Replacing null values with zeros
new_df.fillna(0)
new_df[new_df.fid != '31229']
new_df.dropna()

fid       school_name  status  ... location constituency    county
0      10001.0    CHEBIREBEI PRI  PUBLIC  ...  MOGOTIO      SINENDE   BARINGO
1      10002.0       WASEGES PRI  PUBLIC  ...  MOGOTIO      SINENDE   BARINGO
2      10003.0       KIRIBOT PRI  PUBLIC  ...  MOGOTIO      KABUSWO   BARINGO
3      10004.0   KABUSWO PRI SCH  PUBLIC  ...  MOGOTIO      KABUSWO   BARINGO
4      10006.0         POMBO PRI  PUBLIC  ...  MOGOTIO    NGENDALEL   BARINGO
...        ...               ...     ...  ...      ...          ...       ...
31224  29444.0          URAN PRI  PUBLIC  ...     URAN       MOYALE  MARSABIT
31225  29447.0    URAN GODHA PRI  PUBLIC  ...     URAN       MOYALE  MARSABIT
31226  29451.0         WALDA PRI  PUBLIC  ...    WALDA       MOYALE  MARSABIT
31227  29455.0  DADACH ELELE PRI  PUBLIC  ...    WALDA       MOYALE  MARSABIT
31228  29458.0        RAWANA PRI  PUBLIC  ...   RAWANA       MOYALE  MARSABIT

[30916 rows x 18 columns]

# Data Analysis

Analysis for this project will use the WHO recommended ratio of 1:27 which means one toilet to 27 students per school.

# Status of Sanitation in Provinces, Counties and District as per Student to Toilet Ratio



### What is the status of sanitation in the provinces as per Student to Toilet Ratio

In [ ]:
# The ratio of students to toilets in each province.
prov_toilet=new_df.groupby('province')[['total_toilets']].sum()
prov_toilet
prov_stu=new_df.groupby('province')[['total_students']].sum()
prov_stu
prov_toilet_ratio=((prov_stu['total_students'])/(prov_toilet['total_toilets']))
print(prov_toilet_ratio)


province
CENTRAL          19.840506
COAST            53.638959
EASTERN          28.248324
NAIROBI          33.007191
NORTH EASTERN    53.557662
NYANZA           40.889606
RIFT VALLEY      33.673576
WESTERN          40.230306
dtype: float64


In [ ]:
# Provinces which do not meet WHO's threshold.
prov_toilet_ratio.loc[prov_toilet_ratio > 27.0]

province
COAST            53.638959
EASTERN          28.248324
NAIROBI          33.007191
NORTH EASTERN    53.557662
NYANZA           40.889606
RIFT VALLEY      33.673576
WESTERN          40.230306
dtype: float64

7 out of 8 provinces do not meet WHO's threshold.

### What is the status of sanitation in the various counties as per Student to Toilet Ratio

In [ ]:
# The ratio of students to toilets in each county.
county_toilet=new_df.groupby('county')[['total_toilets']].sum()
county_toilet
county_stu=new_df.groupby('county')[['total_students']].sum()
county_stu
county_toilet_ratio=((county_stu['total_students'])/(county_toilet['total_toilets']))
print(county_toilet_ratio)

county
BARINGO            34.108600
BUNGOMA            44.334590
BUSIA              50.923115
ELGEYO MARAKWET    29.395833
EMBU               20.591123
GARISSA            47.466049
HOMA BAY           50.906281
ISIOLO             36.869231
KAJIADO            33.300576
KAKAMEGA           38.979560
KERICHO            33.999337
KIAMBU             22.199413
KILIFI             67.134352
KIRINYAGA          18.606933
KISII              31.890742
KISUMU             48.432141
KITUI              34.758556
KWALE              63.232720
LAIKIPIA           23.531505
LAMU               48.496386
MACHAKOS           30.258517
MAKUENI            31.521811
MALINDI            72.488428
MANDERA            58.126404
MARSABIT           41.610000
MERU               23.588294
MIGORI             52.071533
MOMBASA            39.397296
MURANGA            17.086188
NAIROBI            33.007191
NAKURU             30.627977
NANDI              27.854085
NAROK              42.434348
NYAMIRA            26.317897
NYANDAR

In [ ]:
# Counties which do not meet WHO's threshold.
print('Counties that do not meet WHO threshold are',county_toilet_ratio.loc[county_toilet_ratio>27.0].count())
county_toilet_ratio.loc[county_toilet_ratio > 27.0]


Counties that do not meet WHO threshold are 37


county
BARINGO            34.108600
BUNGOMA            44.334590
BUSIA              50.923115
ELGEYO MARAKWET    29.395833
GARISSA            47.466049
HOMA BAY           50.906281
ISIOLO             36.869231
KAJIADO            33.300576
KAKAMEGA           38.979560
KERICHO            33.999337
KILIFI             67.134352
KISII              31.890742
KISUMU             48.432141
KITUI              34.758556
KWALE              63.232720
LAMU               48.496386
MACHAKOS           30.258517
MAKUENI            31.521811
MALINDI            72.488428
MANDERA            58.126404
MARSABIT           41.610000
MIGORI             52.071533
MOMBASA            39.397296
NAIROBI            33.007191
NAKURU             30.627977
NANDI              27.854085
NAROK              42.434348
SAMBURU            35.615385
SIAYA              44.806595
TAITA TAVETA       33.412991
TANA RIVER         64.356667
TRANS NZOIA        44.022541
TURKANA            68.597725
UASIN GISHU        30.390635
VIHIGA 

### What is the status of sanitation in districts as per Student to Toilet 

In [ ]:
# The ratio of students to toilets in each district.
district_toilet=new_df.groupby('district')[['total_toilets']].sum()
district_toilet
district_stu=new_df.groupby('district')[['total_students']].sum()
district_stu
district_toilet_ratio=((district_stu['total_students'])/(district_toilet['total_toilets']))
district_toilet_ratio.tail()

district
TURKANA        68.597725
UASIN GISHU    30.390635
VIHIGA         28.939490
WAJIR          54.846011
WEST POKOT     43.479452
dtype: float64

In [ ]:
# Districts which do not meet WHO's threshold.
print('Districts that do not meet WHO threshold are',district_toilet_ratio.loc[district_toilet_ratio>27.0].count())
district_toilet_ratio.loc[district_toilet_ratio > 27.0]

Districts that do not meet WHO threshold are 56


district
BARINGO          35.067899
BOMET            35.261664
BONDO            44.559932
BUNGOMA          46.664319
BURETI           36.578166
BUSIA            51.340030
BUTERE/MUMIAS    43.677281
CENTRAL KISII    31.561610
GARISSA          46.091082
GUCHA            32.205228
HOMA BAY         64.559946
IJARA            53.454545
ISIOLO           36.869231
KAJIADO          33.300576
KAKAMEGA         39.177331
KERICHO          32.925532
KILIFI           67.134352
KISUMU           47.492621
KITUI            32.686863
KOIBATEK         32.641919
KURIA            47.167987
KWALE            63.232720
LAMU             48.496386
LUGARI           31.935227
MACHAKOS         30.258517
MAKUENI          31.521811
MALINDI          72.488428
MANDERA          58.126404
MARAKWET         34.201044
MARSABIT         36.815113
MERU NORTH       32.809804
MIGORI           53.837090
MOMBASA          39.397296
MOYALE           49.500000
MT ELGON         30.035073
MWINGI           38.392265
NAIROBI          33

# Status of sanitation in Kenyan Primary schools

In [ ]:
# The number of schools with sanitation facilities for boys and girls
sanitation = new_df.loc[(new_df['toilet_ratio']<= 27.0)]
sanitation.nsmallest(50, 'toilet_ratio')
print("The number of Kenyan primary schools that meet WHO threshold =", sanitation['fid'].count())

# calculating percentage of schools that meet WHO threshold of 1:27
percentage =100 *((sanitation['fid'].count())/(prim_1['fid'].count()))
print("The percentage of Kenyan Primary schools that meet WHO threshold =",percentage,"%")

The number of Kenyan primary schools that meet WHO threshold = 15077
The percentage of Kenyan Primary schools that meet WHO threshold = 48.27884338275321 %


48.3% of Kenyan Primary Schools meet the WHO thresholds.

In [ ]:
# The number of schools with sanitation facilities for boys and girls
sanitation_not = new_df.loc[(new_df['toilet_ratio']> 27.0)]
# sanitation_not.nsmallest(, 'toilet_ratio')
print("The number of Kenyan primary schools that do not meet WHO threshold =", sanitation_not['fid'].count())

# calculating percentage of schools that do not meet WHO threshold of 1:27
percentage_not =100 *((sanitation_not['fid'].count())/(prim_1['fid'].count()))
print("The percentage of Kenyan Primary schools that do meet WHO threshold =",percentage_not,"%")

The number of Kenyan primary schools that do not meet WHO threshold = 16152
The percentage of Kenyan Primary schools that do meet WHO threshold = 51.72115661724679 %


51.7% of Kenyan primary schools do not meet WHO threshold.

In [ ]:
# The list of schools that don't meet the WHO threshold.
sanitation_not

fid      school_name  status  ... location constituency    county
1      10002.0      WASEGES PRI  PUBLIC  ...  MOGOTIO      SINENDE   BARINGO
2      10003.0      KIRIBOT PRI  PUBLIC  ...  MOGOTIO      KABUSWO   BARINGO
3      10004.0  KABUSWO PRI SCH  PUBLIC  ...  MOGOTIO      KABUSWO   BARINGO
4      10006.0        POMBO PRI  PUBLIC  ...  MOGOTIO    NGENDALEL   BARINGO
5      10007.0  KIPNYUNGUNY PRI  PUBLIC  ...  MOGOTIO    NGENDALEL   BARINGO
...        ...              ...     ...  ...      ...          ...       ...
31222  29437.0    KARBURURI PRI  PUBLIC  ...   GALOLE       MOYALE  MARSABIT
31223  29440.0       GOLOLE PRI  PUBLIC  ...   GALOLE       MOYALE  MARSABIT
31224  29444.0         URAN PRI  PUBLIC  ...     URAN       MOYALE  MARSABIT
31225  29447.0   URAN GODHA PRI  PUBLIC  ...     URAN       MOYALE  MARSABIT
31226  29451.0        WALDA PRI  PUBLIC  ...    WALDA       MOYALE  MARSABIT

[16152 rows x 18 columns]

In [ ]:
# here is the number of schools in each province that do not meet the WHO threshold
sanitation_not_prov=sanitation_not.groupby(['province'])[['school_name']].count()
sanitation_not_prov.

school_name
province                  
CENTRAL                645
COAST                 1123
EASTERN               4778
NAIROBI                450
NORTH EASTERN          229
NYANZA                3245
RIFT VALLEY           3922
WESTERN               1760

In [ ]:
# here is a list of schools for each county in central that do not meet WHO threshold
central_sanitation_not=sanitation_not.loc[sanitation_not['province']=='CENTRAL']
central_sanitation_not.set_index('county')

fid                       school_name  ...    location constituency
county                                             ...                         
KIAMBU   6113.0                     KINOO PRIMARY  ...       KINOO       KABETE
KIAMBU   6122.0         SATHYA SAI PRIMARY SCHOOL  ...       KINOO       KABETE
KIAMBU   6128.0             UTHIRU PRIMARY SCHOOL  ...       KINOO       KABETE
KIAMBU   6134.0          ACACIA GREEN ACADEMY PRI  ...       KINOO       KABETE
KIAMBU   6140.0  MAMA NGINA KENYATTA PCEA PRIMARY  ...       KINOO       KABETE
...         ...                               ...  ...         ...          ...
MURANGA  5642.0                    NDIKWE PRI SCH  ...       MBIRI       KIHARU
MURANGA  5648.0                 GITHUGUYA PRI SCH  ...       MBIRI       KIHARU
MURANGA  5807.0                   KAGANJO PRI SCH  ...  KAMACHARIA     MATHIOYA
MURANGA  5861.0                    KIRIKO PRI SCH  ...        KIRU     MATHIOYA
MURANGA  5894.0                     UMBUI PRI SCH  ...        KIRU     MATHIOYA

[645 rows x 17 columns]

In [ ]:
# here is a list of schools for each county in coast that do not meet WHO threshold
coast_sanitation_not=sanitation_not.loc[sanitation_not['province']=='COAST']
coast_sanitation_not.set_index('county')

fid                         school_name  ...     location constituency
county                                              ...                          
KWALE   2001.0                        VUGA PRI SCH  ...       TSIMBA       MATUGA
KWALE   2003.0  MWANGAZA VISIONARY ACADEMY PRI SCH  ...       TSIMBA       MATUGA
KWALE   2009.0                      VORONI PRI SCH  ...          WAA       MATUGA
KWALE   2012.0                    KOMBANI PRI  SCH  ...          WAA       MATUGA
KWALE   2015.0                      MATUGA PRI SCH  ...          WAA       MATUGA
...        ...                                 ...  ...          ...          ...
LAMU    1708.0                        WITU PRI SCH  ...         WITU    LAMU WEST
LAMU    1715.0                      SOROKO PRI SCH  ...         WITU    LAMU WEST
LAMU    1719.0                   VIPINGONI PRI SCH  ...         WITU    LAMU WEST
LAMU    1722.0                KATSAKAKAIRU PRI SCH  ...  DIDE WARIDE    LAMU WEST
LAMU    1724.0                         MOA PRI SCH  ...  DIDE WARIDE    LAMU WEST

[1123 rows x 17 columns]

In [ ]:
# here is a list of schools for each county in eastern that do not meet WHO threshold
eastern_sanitation_not=sanitation_not.loc[sanitation_not['province']=='EASTERN']
eastern_sanitation_not.set_index('county')

fid          school_name  status  ...  district  location constituency
county                                          ...                                 
MACHAKOS  10005.0   VONDENI PRI SCHOOL  PUBLIC  ...  MACHAKOS     KIVAA      MASINGA
MACHAKOS     44.0       KIKULE PRI SCH  PUBLIC  ...  MACHAKOS   IKATINI      MASINGA
MACHAKOS     52.0  KWA WANZILU PRI SCH  PUBLIC  ...  MACHAKOS   IKATINI      MASINGA
MACHAKOS     57.0        KYENI PRI SCH  PUBLIC  ...  MACHAKOS  IKALAASA        MWALA
MACHAKOS     58.0   KWAMULINGA PRI SCH  PUBLIC  ...  MACHAKOS   IKATINI      MASINGA
...           ...                  ...     ...  ...       ...       ...          ...
MARSABIT  29437.0        KARBURURI PRI  PUBLIC  ...    MOYALE    GALOLE       MOYALE
MARSABIT  29440.0           GOLOLE PRI  PUBLIC  ...    MOYALE    GALOLE       MOYALE
MARSABIT  29444.0             URAN PRI  PUBLIC  ...    MOYALE      URAN       MOYALE
MARSABIT  29447.0       URAN GODHA PRI  PUBLIC  ...    MOYALE      URAN       MOYALE
MARSABIT  29451.0            WALDA PRI  PUBLIC  ...    MOYALE     WALDA       MOYALE

[4778 rows x 17 columns]

In [ ]:
# here is a list of schools for each county in nairobi that do not meet WHO threshold
nairobi_sanitation_not=sanitation_not.loc[sanitation_not['province']=='NAIROBI']
nairobi_sanitation_not.set_index('county')

fid  ... constituency
county            ...             
NAIROBI  21271.0  ...    DAGORETTI
NAIROBI  21288.0  ...    DAGORETTI
NAIROBI  21294.0  ...    DAGORETTI
NAIROBI  21326.0  ...    DAGORETTI
NAIROBI  21354.0  ...    KAMUKUNJI
...          ...  ...          ...
NAIROBI  22963.0  ...     KASARANI
NAIROBI  22965.0  ...     KASARANI
NAIROBI  22971.0  ...     KASARANI
NAIROBI  22975.0  ...     KASARANI
NAIROBI  22993.0  ...      STAREHE

[450 rows x 17 columns]

In [ ]:
# here is a list of schools for each county in north eastern that do not meet WHO threshold
northeastern_sanitation_not=sanitation_not.loc[sanitation_not['province']=='NORTH EASTERN']
northeastern_sanitation_not.set_index('county')

fid             school_name  ...   location     constituency
county                                    ...                            
MANDERA  21011.0       RHAMU DEB PRI SCH  ...      RHAMU  MANDERA CENTRAL
MANDERA  21013.0       AL FOWZAN PRI SCH  ...     GIRISA  MANDERA CENTRAL
MANDERA  21014.0  RHAMU ARIDZONE PRI SCH  ...     GIRISA  MANDERA CENTRAL
MANDERA  21016.0       SHANTOLEY PRI SCH  ...  SHANTOLEY  MANDERA CENTRAL
MANDERA  21018.0       AL-HIDAYA PRI SCH  ...      RHAMU  MANDERA CENTRAL
...          ...                     ...  ...        ...              ...
WAJIR    21889.0        BUTEHELU PRI SCH  ...     BATALU      WAJIR NORTH
WAJIR    21892.0           AJAWA PRI SCH  ...       AJAW      WAJIR NORTH
WAJIR    21895.0          QUDAMA PRI SCH  ...     QUDAMA      WAJIR NORTH
WAJIR    21899.0          DANABA PRI SCH  ...     DANABA      WAJIR NORTH
WAJIR    21904.0           GURAR PRI SCH  ...      GURAR      WAJIR NORTH

[229 rows x 17 columns]

In [ ]:
# here is a list of schools for each county in nyanza that do not meet WHO threshold
nyanza_not=sanitation_not.loc[sanitation_not['province']=='NYANZA']
nyanza_sanitation_not.set_index('county')

fid  ...      constituency
county           ...                  
MIGORI   7703.0  ...            MIGORI
MIGORI  16004.0  ...           NYATIKE
MIGORI  16006.0  ...           NYATIKE
MIGORI  16007.0  ...           NYATIKE
MIGORI  16010.0  ...           NYATIKE
...         ...  ...               ...
KISII   19860.0  ...     KITUTU CHACHE
KISII   19861.0  ...     KITUTU CHACHE
KISII   19862.0  ...     KITUTU CHACHE
KISII   19866.0  ...  NYARIBARI CHACHE
KISII   19869.0  ...  NYARIBARI CHACHE

[3245 rows x 18 columns]

In [ ]:
# here is a list of schools for each county in rift valley that do not meet WHO threshold
rfvalley_sanitation_not=sanitation_not.loc[sanitation_not['province']=='RIFT VALLEY']
rfvalley_sanitation_not.set_index('county')

fid                    school_name  ...       location constituency
county                                           ...                            
BARINGO  10002.0                    WASEGES PRI  ...        MOGOTIO      SINENDE
BARINGO  10003.0                    KIRIBOT PRI  ...        MOGOTIO      KABUSWO
BARINGO  10004.0                KABUSWO PRI SCH  ...        MOGOTIO      KABUSWO
BARINGO  10006.0                      POMBO PRI  ...        MOGOTIO    NGENDALEL
BARINGO  10007.0                KIPNYUNGUNY PRI  ...        MOGOTIO    NGENDALEL
...          ...                            ...  ...            ...          ...
TURKANA   8996.0     COSMAS NAPOPONGOIT PRI SCH  ...  TURKANA NORTH       MOGILA
TURKANA   8997.0        GLOBAL SEED ACADEMY PRI  ...  TURKANA NORTH  LOKICHOGGIO
TURKANA   8998.0             LOKI LIGHT ACADEMY  ...  TURKANA NORTH  LOKICHOGGIO
TURKANA   8999.0  AIC LOKICHOGGIO MIXED PRI SCH  ...  TURKANA NORTH  LOKICHOGGIO
TURKANA   9000.0      KANUKURUDIO MIXED PRI SCH  ...  TURKANA NORTH       KAERIS

[3922 rows x 17 columns]

In [ ]:
# here is a list of schools for each county in western that do not meet WHO threshold
western_sanitation_not=sanitation_not.loc[sanitation_not['province']=='WESTERN']
western_sanitation_not.set_index('county')

fid                   school_name  ...   location constituency
county                                          ...                        
BUNGOMA  13031.0                NYANGE PAG PRI  ...   TONGAREN     KIMILILI
BUNGOMA  13035.0            SIANGALAMWE RC PRI  ...   TONGAREN     KIMILILI
BUNGOMA  13039.0                KAKAMWE RC PRI  ...   TONGAREN     KIMILILI
BUNGOMA  13045.0              TONGAREN DEB PRI  ...   TONGAREN     KIMILILI
BUNGOMA  13047.0               BINYENYA RC PRI  ...   TONGAREN     KIMILILI
...          ...                           ...  ...        ...          ...
BUNGOMA  16336.0            CHESIRO SA PRI SCH  ...   CHEPTAIS     MT ELGON
BUNGOMA  16340.0          PERESTEN ACK PRI SCH  ...   CHEPTAIS     MT ELGON
BUNGOMA  16346.0  CHEPTAIS SDA ACADEMY PRI SCH  ...   CHEPTAIS     MT ELGON
BUNGOMA  16356.0                KIMAMA PRI SCH  ...  CHESIKAKI     MT ELGON
BUNGOMA  16360.0              CHESIYWO PRI SCH  ...  CHESIKAKI     MT ELGON

[1760 rows x 17 columns]

# Schools in Kenya without sanitation facilities

In [ ]:
# The number of schools without sanitation facilities for girls, boys and teachers
no_sanitation = new_df.loc[(new_df['boystoilet']==0.0)&(new_df['girlstoilet']==0.0)&(new_df['teachers_toilet']==0.0)]
print("The number of schools without sanitation =",no_sanitation['fid'].count())
print("The list of schools without sanitation is shown below")
no_sanitation

The number of schools without sanitation = 2151
The list of schools without sanitation is shown below


fid                school_name  ... constituency    county
101    10391.0           MUMBERES PRI SCH  ...     MUMBERES   BARINGO
166    10648.0         KAPKARARAM PRI SCH  ...    NGUBERETI   BARINGO
193    10821.0   REHOBOTH ACADEMY PRI SCH  ...     KATHIANI  MACHAKOS
223       77.0       HAVENTON ACA PRI SCH  ...      MASINGA  MACHAKOS
238      101.0             EENDEI PRI SCH  ...      MASINGA  MACHAKOS
...        ...                        ...  ...          ...       ...
31195  29324.0  MOYALE SOMENI ACADEMY PRI  ...       MOYALE  MARSABIT
31214  29409.0        SOLOLO SPECIAL UNIT  ...       MOYALE  MARSABIT
31216  29416.0                  ANONA PRI  ...       MOYALE  MARSABIT
31227  29455.0           DADACH ELELE PRI  ...       MOYALE  MARSABIT
31228  29458.0                 RAWANA PRI  ...       MOYALE  MARSABIT

[2151 rows x 18 columns]

In [ ]:
# Number of schools that did not have boys toilets
no_boys_sanitation = new_df.loc[(new_df['boystoilet']==0.0)]
print("The number of schoos without sanitation for boys =",no_boys_sanitation['fid'].count())
print("The list of schools without sanitation for boys is shown below")
no_boys_sanitation

The number of schoos without sanitation for boys = 2648
The list of schools without sanitation for boys is shown below


fid               school_name  ...    constituency    county
53     10105.0           METIPSO PRI SCH  ...  LEMBUS CENTRAL   BARINGO
101    10391.0          MUMBERES PRI SCH  ...        MUMBERES   BARINGO
166    10648.0        KAPKARARAM PRI SCH  ...       NGUBERETI   BARINGO
193    10821.0  REHOBOTH ACADEMY PRI SCH  ...        KATHIANI  MACHAKOS
205       54.0         KITONYINI PRIMARY  ...   MACHAKOS TOWN  MACHAKOS
...        ...                       ...  ...             ...       ...
31214  29409.0       SOLOLO SPECIAL UNIT  ...          MOYALE  MARSABIT
31215  29413.0               HAWECHA PRI  ...          MOYALE  MARSABIT
31216  29416.0                 ANONA PRI  ...          MOYALE  MARSABIT
31227  29455.0          DADACH ELELE PRI  ...          MOYALE  MARSABIT
31228  29458.0                RAWANA PRI  ...          MOYALE  MARSABIT

[2648 rows x 18 columns]

In [ ]:
# Number of schools that did not have girls toilets
no_girls_sanitation = new_df.loc[(new_df['girlstoilet']==0.0)]
print("The number of schools without sanitation for girls =",no_girls_sanitation['fid'].count())
print("The list of schools without sanitation for girls is shown below")
no_girls_sanitation

The number of schoos without sanitation for girls = 2564
The list of schools without sanitation for girls is shown below


fid                school_name  ...    constituency    county
53     10105.0            METIPSO PRI SCH  ...  LEMBUS CENTRAL   BARINGO
56     10116.0                  ARAMA PRI  ...  LEMBUS CENTRAL   BARINGO
101    10391.0           MUMBERES PRI SCH  ...        MUMBERES   BARINGO
139    10541.0   MAJI MAZURI SPECIAL UNIT  ...     MAJI MAZURI   BARINGO
166    10648.0         KAPKARARAM PRI SCH  ...       NGUBERETI   BARINGO
...        ...                        ...  ...             ...       ...
31195  29324.0  MOYALE SOMENI ACADEMY PRI  ...          MOYALE  MARSABIT
31214  29409.0        SOLOLO SPECIAL UNIT  ...          MOYALE  MARSABIT
31216  29416.0                  ANONA PRI  ...          MOYALE  MARSABIT
31227  29455.0           DADACH ELELE PRI  ...          MOYALE  MARSABIT
31228  29458.0                 RAWANA PRI  ...          MOYALE  MARSABIT

[2564 rows x 18 columns]

In [ ]:
# Number of schools that did not have teachers toilets
no_teachers_sanitation = new_df.loc[(new_df['teachers_toilet']==0.0)]
print("The number of schoos without sanitation for teachers =",no_teachers_sanitation['fid'].count())
print("The list of schools without sanitation for teachers is shown below")
no_teachers_sanitation

The number of schoos without sanitation for teachers = 3127
The list of schools without sanitation for teachers is shown below


fid               school_name  ...    constituency    county
6      10008.0              KABERGEI PRI  ...         OLKOKWE   BARINGO
51     10098.0  EMMANUEL ACADEMY PRI SCH  ...        MUMBERES   BARINGO
56     10116.0                 ARAMA PRI  ...  LEMBUS CENTRAL   BARINGO
73     10312.0      HERO ACADEMY PRI SCH  ...      KOIBOS SOI   BARINGO
101    10391.0          MUMBERES PRI SCH  ...        MUMBERES   BARINGO
...        ...                       ...  ...             ...       ...
31216  29416.0                 ANONA PRI  ...          MOYALE  MARSABIT
31219  29426.0                AMBALO PRI  ...          MOYALE  MARSABIT
31221  29434.0          MUKH - GURRA PRI  ...          MOYALE  MARSABIT
31227  29455.0          DADACH ELELE PRI  ...          MOYALE  MARSABIT
31228  29458.0                RAWANA PRI  ...          MOYALE  MARSABIT

[3127 rows x 18 columns]

### Comparison of status of sanitation between public and private schools

In [ ]:
# Status of sanitation in Public Schools
public =new_df.loc[new_df['status']=='PUBLIC']
public
print("Total number of public schools =",public['fid'].count())

# Calculate total number of public schools without sanitation facilities
pub_1 = public.loc[(public['boystoilet']==0.0)&(public['girlstoilet']==0.0)&(public['teachers_toilet']==0.0)]
print("Total number of public schools without sanitation facilities =",pub_1['fid'].count())

# Calculating ratio of public shools to public schools without sanitation facilities
pub_ratio = float((pub_1['fid'].count())/(public['fid'].count()))
print("Percentage of public schools without sanitation =",(pub_ratio*100.0),"%")
print("The list of public schools without sanitation is shown below")
pub_1

Total number of public schools = 23792
Total number of public schools without sanitation facilities = 1119
Percentage of public schools without sanitation = 4.703261600537996 %
The list of public schools without sanitation is shown below


fid          school_name  ... constituency    county
101    10391.0     MUMBERES PRI SCH  ...     MUMBERES   BARINGO
166    10648.0   KAPKARARAM PRI SCH  ...    NGUBERETI   BARINGO
238      101.0       EENDEI PRI SCH  ...      MASINGA  MACHAKOS
269      155.0      MILAANI PRI SCH  ...      MASINGA  MACHAKOS
325      228.0   TANA RANCH PRI SCH  ...      MASINGA  MACHAKOS
...        ...                  ...  ...          ...       ...
31182  29275.0  BUTIYE SPECIAL UNIT  ...       MOYALE  MARSABIT
31214  29409.0  SOLOLO SPECIAL UNIT  ...       MOYALE  MARSABIT
31216  29416.0            ANONA PRI  ...       MOYALE  MARSABIT
31227  29455.0     DADACH ELELE PRI  ...       MOYALE  MARSABIT
31228  29458.0           RAWANA PRI  ...       MOYALE  MARSABIT

[1119 rows x 18 columns]

4.7% of public schools did not have sanitation facilities.

In [ ]:
# Status of sanitation in Private Schools
private = new_df.loc[new_df['status']=='PRIVATE']
private
print("Total number of private schools =",private['fid'].count())

# Calculate total number of public schools without sanitation facilities
priv_1 = private.loc[(private['boystoilet']==0.0)&(private['girlstoilet']==0.0)&(private['teachers_toilet']==0.0)]
priv_1['fid'].count()
print ("Total number of private schools without sanitation facilities =",priv_1['fid'].count())

# Calculating ratio of private schools to private schools without sanitation facilities
priv_ratio = float((priv_1['fid'].count())/(private['fid'].count()))
print("Percentage of private schools without sanitation =",(priv_ratio*100.0),"%")
print("The list of private schools without sanitation is shown below")
priv_1

Total number of private schools = 7437
Total number of private schools without sanitation facilities = 1032
Percentage of private schools without sanitation = 13.876563130294473 %
The list of private schools without sanitation is shown below


fid                      school_name  ... constituency    county
193    10821.0         REHOBOTH ACADEMY PRI SCH  ...     KATHIANI  MACHAKOS
223       77.0             HAVENTON ACA PRI SCH  ...      MASINGA  MACHAKOS
329      267.0                ST JOSEPH PRIMARY  ...     KANGUNDO  MACHAKOS
345      355.0                  JOY ACA PRI SCH  ...      MASINGA  MACHAKOS
364      385.0  NDOVOINI HOME FOR PHY. DISABLED  ...      MASINGA  MACHAKOS
...        ...                              ...  ...          ...       ...
30616  28818.0              SA GICHICHE PRI SCH  ...    RUNYENJES      EMBU
30675  23653.0                    MERCY PRY SCH  ...     LAISAMIS  MARSABIT
30745  28679.0                    MERCY PRY SCH  ...     LAISAMIS  MARSABIT
31134  24125.0        MOYALE SOMENI ACADEMY PRI  ...       MOYALE  MARSABIT
31195  29324.0        MOYALE SOMENI ACADEMY PRI  ...       MOYALE  MARSABIT

[1032 rows x 18 columns]

13.9% of private schools did not have sanitation facilities.

4.7% of public schools lacked sanitation facilities, while 13.9% of private schools did not have sanitation facilities.

### Comparison of status of sanitation between day and boarding schools

In [ ]:
# Status of sanitation in day schools
# Calculating the total number of day schools.
day = new_df.loc[(new_df['type2']=='DAY ONLY')]
print("Total number of day schools =", day['fid'].count())

# Calculating the number of day schools without sanitation.
day_1 = day.loc[(day['boystoilet']==0.0)&(day['girlstoilet']==0.0)&(day['teachers_toilet']==0.0)]
print("Total number of day primary schools without sanitation =", day_1['fid'].count())

# Calculating percentage of day primary schools without sanitation facilities.
day_ratio = float((day_1['fid'].count())/(day['fid'].count()))
print("Percentage of day schools without sanitation =",(day_ratio*100.0),"%")

Total number of day schools = 28572
Total number of day primary schools without sanitation = 1725
Percentage of day schools without sanitation = 6.037379252414952 %


6.03% of Kenyan day schools did not have sanitation facilities

In [ ]:
# Status of sanitation in day/boarding schools
# Calculating the total number of day/boarding schools.
dayboard = new_df.loc[(new_df['type2']=='DAY & BOARDING')]
print("Total number of day/boarding schools =", dayboard['fid'].count())

# Calculating the number of day/boarding schools without sanitation.
dayboard_1 = dayboard.loc[(dayboard['boystoilet']==0.0)&(dayboard['girlstoilet']==0.0)&(dayboard['teachers_toilet']==0.0)]
print("Total number of day/boarding primary schools without sanitation =", dayboard_1['fid'].count())

# Calculating percentage of day/boarding primary schools without sanitation facilities.
dayboard_ratio = float((dayboard_1['fid'].count())/(dayboard['fid'].count()))
print("Percentage of day/boarding schools without sanitation =",(dayboard_ratio*100.0),"%")

Total number of day/boarding schools = 1660
Total number of day/boarding primary schools without sanitation = 54
Percentage of day/boarding schools without sanitation = 3.2530120481927707 %


3.25% of Kenyan boarding schools did not have sanitation facilities.

### Which gender is most affected by lack of sanitation facilities?

In [ ]:
# Number of schools that did not have boys toilets
no_boys = new_df.loc[(new_df['boystoilet']==0.0)]
print("The number of schools without boys toilets =", no_boys['fid'].count())

# Number of schools that did not have girls toilets
no_girls = new_df.loc[(new_df['girlstoilet']==0.0)]
print("The number of schools without girls toilets =", no_girls['fid'].count())

The number of schools without boys toilets = 2648
The number of schools without girls toilets = 2564


The most affected gender is male having more schools without toilets for boys

# Conclusion

From our analysis we have answered our research objective. The results can be summarized as follows:

Primary schools in 7 out of 8 provinces in Kenya do not meet WHO threshold.

Primary schools in 56 districts do not meet the WHO threshold.

The number of Kenyan primary schools that meet WHO threshold is 15,077 this is 48.3% of the total number of primary schools.

16,152 Kenyan primary schools do not meet WHO threshold, this is 51.7% of all the primary schools.

The number of schools that do not have boys toilets is 2,648, while the number of schools that do not have girls toilets is 2,564. This shows that more schools lack toilets for boys.

The number of schools without any sanitation facilities in Kenya is 2,151 schools.

From this analysis we have made recommendations to the government to liaise with the relevant bodies to build or improve sanitation facilities where necessary. 
